### DISCUSSION TITLE
Suckless

### DISCUSSION THREAD

**Discussion by pazos** [koreader/koreader#7215] [Ideas]
I'm taking the name from https://suckless.org/philosophy/, but adapting the idea to KOReader.

My proposal would be reducing a bit the features that can be better implemented elsewhere. I barely use most of these features, so I cannot justify why there're part of the program. Others might disagree or have specific workflows that justify these features.

Generally speaking I would avoid features that depend on 3rd party services ***unless*** these services can be self-hosted, like wallabag, joplin or calibre, or they are actually integrated in the core and have stable APIs, like dropbox or wikipedia.

1. remove `evernote.koplugin`

The ability of exporting highlights and annotations is covered by https://github.com/noembryo/KoHighlights. IMHO makes sense to have a general exporter as part of the core (might be blocked by https://github.com/koreader/koreader/issues/5794) that's able to convert the internal representation in a common format that others plugins can repurpose.

2. remove `zsync.koplugin`

My suggestion would be https://github.com/koreader/koreader/issues/7208#issuecomment-769922420

3. remove `send2ebook.koplugin`

It is just a glorified ftp client, which a companion app that doesn't work and was archived a while ago https://github.com/koreader/android-send2ebook

As a FTP client could make sense to keep as a cloud storage option, but again it is covered by other external options, like `rclone`.

4. remove the `backgroundrunner.koplugin`.

It is not a final user plugin. The background plugin should be enabled automatically if one or more plugins that rely on it are enabled.

**Comment by NiLuJe** [comment]
Sounds good to me ;).

4. is mostly already done since @poire-z tamed it a bit, IIRC.

**Reply by poire-z** [reply]
I'm not friend with the hideous BackgroundRunner. I made him shut down when it's not used, but I would be fine just dropping it :)
It's still used for auto frontlight and warmth by:
```
frontend/device/cervantes/powerd.lua
frontend/device/kobo/powerd.lua
plugins/autofrontlight.koplugin/main.lua
```
This could be solved by just using re/scheduleIn (like done these last days for header and footer auto refreshes). But these are "frontlight" stuff, that nobody really wants to touch :)

**Reply by pazos** [reply]
That's a good excuse to kill the autofrontlight feature ^^

**Reply by pazos** [reply]
I was meant the autowarmth feature. The autofrontlight is a kindle thing that I'm not sure how to tackle :/

**Reply by pazos** [reply]
I wrote a PR to disable autowarmth for cervantes in https://github.com/koreader/koreader/pull/7220, but since the counterpart for Kobo won't be removed I see no point to remove the feature for cervantes yet. So closing that one :(

**Comment by NiLuJe** [comment]
In the same vein, this is why I'm unhappy with the ssh plug-in : it's full of platform specific quirks that require annoying to update patches, best implemented in actual ssh tools by people familiar with each platform.

**Reply by pazos** [reply]
Feel free to deprecate it. It is one of these I actually use from time to time, but I can live without the binary bundled in KOReader. The UI part of the plugin is a nice thing to keep IMO, just check `which dropbear`  or something like that at runtime.

**Comment by poire-z** [comment]
No real opinion.
I don't know much about these plugins or how they work - and if people are really using some of I them.
I guess some may suck or even not work, and some are abandonned (which is expected for dedicated use case plugins...) so it may account for 10-20% of our issues - but sometimes, someone pops up with some skills and can fix them (and may disappear just after :/)...
I usually like things tidy, so I would not mind getting rid of some - but as I can and have disabled them, I don't mind them.

It's indeed more about public image and advertising crappy features that none of us is able/willing to support.
So, may be just make them more discreet, like you make some of them "deprecated", mark some as "abandonned" and disabled by default - so newcomers cound think about adopting them :)

**Reply by NiLuJe** [reply]
It's true that there might be a spectrum, ranging from "this might be useful, but is in dire need of a maintainer" (=> flag it deprecated) to "this makes no sense and is generally broken, please kill it with fire", with a "generally a PITA from a maintenance perspective" in the middle ;).

On a practical note, if all the crazy stuff that depends on ridiculously old versions of the øMQ ecosystem is gone, that makes updating to sane versions much easier, which could be beneficial to some things (I'm thinking of the Calibre plugin, which currently can't detect when Calibre closes the socket in a sane way because of a terrible poll wrapper that doesn't handle POLLHUP).

In the same vein, the evernote plugin drags a lot of crazy stuff into the buildsystem.

**Reply by Frenzie** [reply]
> So, may be just make them more discreet, like you make some of them "deprecated", mark some as "abandonned" and disabled by default - so newcomers cound think about adopting them :)

Indeed, ideally we'd do it with a bit of a warning process. Something like, each step being one or two release cycles:

1. Deprecated notice.
2. Disabled by default.
3. Removed.

Unless it's severely broken.

That could also motivate users to step in to fix it up.

**Reply by pazos** [reply]
Deprecate `send2ebook` and `zsync`. Evernote plugin is severly broken as is.

**Comment by Galunid** [comment]
1. I think it's a good idea to get rid of evernote/joplin, but it probably would be a good idea to leave the rest of the plugin out of it (json/html export). Evernote is broken either way, Joplin I don't know if anyone uses that. Worst case scenario people can copy the "old" plugin back.
2. If you meant `kosync.koplugin` then that'd be nope. This plugin is a life saver for me and it's absolutely essential (and can be [self hosted](https://github.com/koreader/koreader-sync-server) as far as I know ^^). If you meant `zsync.koplugin` then I'm fine with that.
3. Never used it, but it seems to be wallabag alternative? No opinion on this one.
4. If possible without breaking stuff ^^

**Reply by pazos** [reply]
Yup, `zsync.koplugin`. Corrected in OP

**Reply by NiLuJe** [reply]
The fact that we all subconsciously *knew* that you were talking about the zsync plugin is a pretty telling argument for putting a nail in its coffin ^^.

**Comment by Frenzie** [comment]
Ideally we'd combine it with making it easier to load your own plugins (if necessary). Or set up some metadata about where to find them, or something. Also see https://github.com/FreshRSS/Extensions/blob/8f5df7dc6d75f3a406a37f36faf475c9ae549051/extensions.json

@pazos 
> Generally speaking I would avoid features that depend on 3rd party services unless these services can be self-hosted, like wallabag, joplin or calibre, or they are actually integrated in the core and have stable APIs, like dropbox or wikipedia.

I don't think self-hosting should be a criterion, although it is something that's affected my personal investment in Wallabag over Pocket. Dropbox may be implemented as part of the core, but we might ask whether it actually belongs there. By which I don't mean to drop it, mind you.

@Galunid 
> Joplin I don't know if anyone uses that.

Several people do afaik.

<hr>

On 2-4, sure, drop them. They're unmaintained, although I don't necessarily quite agree with the just use rclone suggestion. :-)

The Joplin plugin, is the suggestion because of the dependencies in base? Because keep in mind we can always drop the Evernote part and keep the Joplin.

**Reply by pazos** [reply]
> Dropbox may be implemented as part of the core, but we might ask whether it actually belongs there. 

> The Joplin plugin, is the suggestion because of the dependencies in base? Because keep in mind we can always drop the Evernote part and keep the Joplin.

Both the cloudstorage and the evernote plugin are examples of hot potatos. Both can be simplified and kept as a core feature but it is a bit time consuming to do that.

At some point we might want to use something like the dispatcher is doing. Allow plugins to register cloud providers or notes exporters and leave the core as simple as possible.

I also like the idea of a single action/button **export notes** that prompts for the format, instead of using a config panel for that.

**Comment by gerroon** [comment]
Joplin is listed under Evernote plugin. Are you going to remove the Joplin support? I hope not because it seems to work and Joplin is not necessarily a 3rd party service, it is a local desktop application running on a pc.

**Comment by mergen3107** [comment]
I have all these plugins in the OP disabled from day one.
SSH, however, is set up, but I haven't really needed it yet on Kindle (thanks to NiLuJe there is native SSH). I just keep it as a backup, I suppose :D

**Comment by NiLuJe** [comment]
IIRC, the plan would be to gut the Evernote plugin of its actually useful and maintainable parts, when the day finally comes to say off with its head ;).

**Reply by pazos** [reply]
it's on my to-do list. I was waiting a bit for https://github.com/koreader/koreader/issues/4951 code proposals to make the inner guts of the exporter part of the core and implement specific exporters as plugins. But maybe we can duplicate the evernote plugin as is, strip the evernote parts and deprecate the old one soon :tm: That should be very easy to do.

The evernote plugin deprecation message could say something like: "this plugin will be removed soon as Evernote features are broken for a while. Please use *nameOfTheNewPlugin* instead.

**Comment by pazos** [comment]
All done ;) 🥳
